### ISSUE TITLE
Can't update functional test refs

### ISSUE THREAD

**Issue by sshane** [pylint-dev/pylint#9688]
### Bug description

```python
The message counter is checked first before the output text, so we never have a chance to write the updated output file.

Need to flip these two statements:

https://github.com/pylint-dev/pylint/blob/86e194307affcb860244278f0c45d4717850601e/pylint/testutils/lint_module_test.py#L256-L261

so this has a chance to run:

https://github.com/pylint-dev/pylint/blob/86e194307affcb860244278f0c45d4717850601e/pylint/testutils/functional/lint_module_output_update.py#L27-L32
```

### Command used

```shell
python tests/test_functional.py -k abstract --update-functional-output
```

### Pylint output

```python
Failure, depending on if you change the refs
```

### Expected behavior

You can update the refs

### Pylint version

```shell
pylint 3.3.0-dev0
astroid 3.2.2
Python 3.11.4 (main, Jul 20 2023, 22:46:38) [GCC 9.4.0]
```

### OS / Environment

Ubuntu 20.04

**Comment by Pierre-Sassoulas** [comment]
It's "by design", writing a new output file is useless (or misleading you into thinking it worked) if the functional test does not output the expected value. That said the functional test is still going to fail the line after, and you're only going to have a temporarily better output file while you fix the issue with the functional test. Feel free to open a MR to change the order :)

**Comment by sshane** [comment]
Ah, maybe I'm confused about the relationship with the functional test and the output ref. I assumed that the output text file was just a reference we compared the pylint output to, without any ignore tags. And the functional test was just to assert certain warnings on certain lines.

**Comment by Pierre-Sassoulas** [comment]
> I assumed that the output text file was a reference we compared the pylint output to, without any ignore tags. 

Yes, but I'm not sure if I understand the "without any ignore tags" part.

> And the functional test was to assert certain warnings on certain lines.

Yes, or the absence of them, if you have unexpected warning the functional test file fail. (And if the functional test fail file updating the output is going to create a wrong output).

**Comment by sshane** [comment]
I just meant running pylint on the whole file, ignoring the warning annotations. Disregard.

**Comment by Pierre-Sassoulas** [comment]
It seems you're already deep in the truth/code so I'm not sure if this part of the doc might help (maybe to get a general idea): https://pylint.readthedocs.io/en/stable/development_guide/contributor_guide/tests/writing_test.html#functional-tests
